improvement(re-ordering of components)

ashwinKumar0505 opened the pull request:
The components can now be reordered using drag and drop.

ashwinKumar0505 committed:
improvement(reordering of components): the components can now be reordered using drag and drop

ashwinKumar0505 committed:
fix(reordering): re-ordering of block and flex display components

ashwinKumar0505 commented:
The dragged component will be re-ordered only when it crosses half of the next component present to it. As the next component can be present on the right or left and also top or bottom. We have to check both the x-axis and y-axis of the next component while dragging. thus while dragging from top to bottom, Not only y-axis middle is checked but also x-axis is checked. Same if we drag component from left to right, Not only x-axis is checked but also y-axis is checked.This leads to contradiction

ashwinKumar0505 committed:
perf(re-ordering): enabled top to bottom re-ordering for now

sanandnarayan commented on src/hooks/useInteractive.ts:
```diff
@@ -7,6 +7,7 @@ import {
   getIsHovered,
 } from '../core/selectors/components'
 import { getShowLayout, getFocusedComponent } from '../core/selectors/app'
+import { useDropComponent } from './useDropComponent'
```
not sure if this is a good idea. useInteractive is a custom hook which makes a component draggable. Not sure if adding useDropComponent hook here is a good idea

ashwinKumar0505 replied:
As useInteractive is a common hook , I used there to reduce the lines of code and also we need ref to know the boundary points of the component.  we are creating ref in the useInteractive. These are the two reasons why i used useDropComponent in useInteractive. I will find a way to use it outside bro

sanandnarayan commented on src/hooks/useDropComponent.ts:
```diff
@@ -1,26 +1,66 @@
+        const hoverClientY = clientOffset && clientOffset.y - top
+
+        // Dragging downwards
+        if (fromIndex < toIndex && hoverClientY && hoverClientY < hoverMiddleY)
```
Need better comments. 
`hasPassedMid?`

sanandnarayan reviewed (changes requested):
You might want to rethink the implementation. 
can we get inspiration from 
`https://react-dnd.github.io/react-dnd/examples/sortable/simple`

ashwinKumar0505 commented:
> You might want to rethink the implementation.
> can we get inspiration from
> `https://react-dnd.github.io/react-dnd/examples/sortable/simple`

This is implemented with the help of that only bro.

ashwinKumar0505 committed:
fix(re-ordering): index issue

While dragging one component to another within same parent , the id of the component parent is also
focused , The handling of the children and parent id is done here .

ashwinKumar0505 committed:
updated the comments

ashwinKumar0505 committed:
refactor(re-ordering): remove of unneccessary additional parameter for useDropComponent

ashwinKumar0505 committed:
perf(re-ordering of components): added useHoverComponent hooks

useHoverComponent hooks are added to handle the ordering of components.\

ashwinKumar0505 committed:
perf(moving of components): not moved if the index of the to-be moved is -1
